core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,268 @@
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+final class DefaultAsyncLoader<T> implements AsyncLoader<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultAsyncLoader.class);
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<
+            DefaultAsyncLoader, RefreshingFuture> loadFutureUpdater = AtomicReferenceFieldUpdater
+            .newUpdater(DefaultAsyncLoader.class, RefreshingFuture.class, "loadFuture");
+    private static final ExecutorService DEFAULT_REFRESH_EXECUTOR = Executors.newSingleThreadExecutor();
+
+    private final Function<@Nullable T, CompletableFuture<T>> loader;
+    @Nullable
+    private final Duration expireAfterLoad;
+    @Nullable
+    private final Predicate<@Nullable T> expireIf;
+    @Nullable
+    private final Predicate<@Nullable T> refreshIf;
+    @Nullable
+    private final ExecutorService refreshExecutor;
+    @Nullable
+    private final BiFunction<Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler;
+
+    private volatile RefreshingFuture<T> loadFuture = RefreshingFuture.completedFuture();
+
+    DefaultAsyncLoader(Function<@Nullable T, CompletableFuture<T>> loader,
+                       @Nullable Duration expireAfterLoad,
+                       @Nullable Predicate<@Nullable T> expireIf,
+                       @Nullable Predicate<@Nullable T> refreshIf,
+                       @Nullable ExecutorService refreshExecutor,
+                       @Nullable BiFunction<
+                               Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler) {
+        requireNonNull(loader, "loader");
+        this.loader = loader;
+        this.expireAfterLoad = expireAfterLoad;
+        this.expireIf = expireIf;
+        this.refreshIf = refreshIf;
+        this.refreshExecutor = refreshExecutor;
+        this.exceptionHandler = exceptionHandler;
+    }
+
+    @Override
+    public CompletableFuture<T> get() {
+        return get0().thenApply(f -> f.loadVal);
+    }
+
+    private CompletableFuture<CacheEntry<T>> get0() {
+        RefreshingFuture<T> future;
+        CacheEntry<T> cacheEntry = null;
+        boolean needsRefresh = false;
+        for (;;) {
+            final RefreshingFuture<T> loadFuture = this.loadFuture;
+            if (!loadFuture.isDone()) {
+                return loadFuture;
+            }
+
+            if (!loadFuture.isCompletedExceptionally()) {
+                cacheEntry = loadFuture.join();
+                if (isValid(cacheEntry)) {
+                    needsRefresh = needsRefresh(cacheEntry);
+                    if (!needsRefresh || loadFuture.refreshing) {
+                        return loadFuture;
+                    }
```

Comment:
<img width="500" alt="image" src="https://github.com/line/armeria/assets/34854527/012ab6db-41d5-4cbc-a28f-1945dae11aba">

I make flow diagram for your understanding 🙇